mcrouter/lib/network/test/TestClientServerUtil.cpp

```diff
@@ -87,6 +87,8 @@ void TestServerOnRequest::onRequest(
     processReply(std::move(ctx), Reply(carbon::Result::NOTFOUND));
   } else if (req.key_ref()->fullKey() == "shutdown") {
     shutdownLock_.post();
+    // TODO
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
I've observed a race condition in several tests in `AsyncMcClientTestSync`, e.g. `qosClass`, where the test would hang on shutdown without this, with the test thread waiting on the `->shutdown` call, the server thread waiting on `->join` and event loop threads still selecting, so I added this as a workaround since I wasn't able to find the exact cause of the race.

Reply:
Some things I've tried to attempt to resolve this:
* reordering the `shutdownLock_` post,
* introducing another synchronization barrier between the call to `shutdown` and `join` on the mock server main thread,
* changing the worker loop to use `loopOnce` over `loop` and check `isShutdown` on each loop iteration, like some of the other mock servers do.